File: src/main/queue/python/exec_queue.py
import time
import sys
from statistics import median

from queue import Queue

class ExecQueue:
    METHOD_NAMES = [
        "addLast",
        "addFirst",
        "add",
        "removeFirst",
        "removeLast",
        "remove",
        "getFirst",
        "getLast",
        "get"
    ]

    @staticmethod
    def main():
        if len(sys.argv) < 2:
            print("Uso: python exec_queue.py <arquivo_entrada>")
            return

        input_file = sys.argv[1]
        output_dir = "data/results/time/"

        try:
            # Cria o diretório de saída se não existir
            import os
            os.makedirs(output_dir, exist_ok=True)

            # Abre os arquivos de saída para escrita
            writers = []
            for method in ExecQueue.METHOD_NAMES:
                writers.append(open(f"{output_dir}{method}.data", "a"))

            with open(input_file, 'r') as reader:
                for line in reader:
                    line = line.strip()
                    if not line:
                        continue

                    elements = list(map(int, line.split()))
                    ExecQueue.test_queue_operations(writers, elements)

        except IOError as e:
            print(f"Erro ao processar arquivos: {e}")
        finally:
            # Fecha todos os arquivos de saída
            for writer in writers:
                writer.close()

    @staticmethod
    def test_queue_operations(writers, elements):
        length = len(elements)
        middle = length // 2
        RUNS = 30
        all_times = [[] for _ in range(len(ExecQueue.METHOD_NAMES))]

        for run in range(RUNS):
            queue = Queue(length + 1)

            # Preenche a fila com os elementos iniciais
            for num in elements:
                queue.add_last(num)

            # Testa addLast
            start_time = time.perf_counter_ns()
            queue.add_last(999)
            end_time = time.perf_counter_ns()
            all_times[0].append(end_time - start_time)
            queue.remove_last()

            # Testa addFirst
            start_time = time.perf_counter_ns()
            queue.add_first(999)
            end_time = time.perf_counter_ns()
            all_times[1].append(end_time - start_time)
            queue.remove_first()

            # Testa add (middle)
            start_time = time.perf_counter_ns()
            queue.add(999, middle)
            end_time = time.perf_counter_ns()
            all_times[2].append(end_time - start_time)
            queue.remove(middle)

            # Testa removeFirst
            start_time = time.perf_counter_ns()
            queue.remove_first()
            end_time = time.perf_counter_ns()
            all_times[3].append(end_time - start_time)
            queue.add_first(elements[0])

            # Testa removeLast
            start_time = time.perf_counter_ns()
            queue.remove_last()
            end_time = time.perf_counter_ns()
            all_times[4].append(end_time - start_time)
            queue.add_last(elements[-1])

            # Testa remove (middle)
            start_time = time.perf_counter_ns()
            queue.remove(middle)
            end_time = time.perf_counter_ns()
            all_times[5].append(end_time - start_time)
            queue.add_last(999)

            # Testa getFirst
            start_time = time.perf_counter_ns()
            queue.get_first()
            end_time = time.perf_counter_ns()
            all_times[6].append(end_time - start_time)

            # Testa getLast
            start_time = time.perf_counter_ns()
            queue.get_last()
            end_time = time.perf_counter_ns()
            all_times[7].append(end_time - start_time)

            # Testa get (middle)
            start_time = time.perf_counter_ns()
            queue.get(middle)
            end_time = time.perf_counter_ns()
            all_times[8].append(end_time - start_time)

        # Calcula a mediana dos tempos e escreve nos arquivos
        for i, method in enumerate(ExecQueue.METHOD_NAMES):
            if all_times[i]:
                median_time = int(median(all_times[i]))
                writers[i].write(f"queue-python {median_time} {length}\n")
                writers[i].flush()

if __name__ == "__main__":
    ExecQueue.main()

File: src/main/queue/cpp/exec_queue.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <chrono>
#include <stdexcept>
#include "queue.cpp"

using namespace std;
using namespace std::chrono;

const vector<string> METHOD_NAMES = {
    "addLast",
    "addFirst",
    "add",
    "removeFirst",
    "removeLast",
    "remove",
    "getFirst",
    "getLast",
    "get"
};

void testQueueOperations(const vector<int>& elements, ofstream writers[]) {
    int length = elements.size();
    int middle = length / 2;
    const int RUNS = 30;
    vector<vector<long>> allTimes(METHOD_NAMES.size(), vector<long>(RUNS));

    for (int run = 0; run < RUNS; run++) {
        Queue<int> queue(length + 3);
        
        // Preenche a fila com os elementos iniciais
        for (int num : elements) {
            queue.addLast(num);
        }

        // Testa addLast
        auto startTime = high_resolution_clock::now();
        queue.addLast(999);
        auto endTime = high_resolution_clock::now();
        allTimes[0][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa addFirst
        startTime = high_resolution_clock::now();
        queue.addFirst(999);
        endTime = high_resolution_clock::now();
        allTimes[1][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa add (middle)
        startTime = high_resolution_clock::now();
        queue.add(999, middle);
        endTime = high_resolution_clock::now();
        allTimes[2][run] = duration_cast<nanoseconds>(endTime - startTime).count();
        
        // Testa removeFirst
        startTime = high_resolution_clock::now();
        queue.removeFirst();
        endTime = high_resolution_clock::now();
        allTimes[3][run] = duration_cast<nanoseconds>(endTime - startTime).count();
        
        // Testa removeLast
        startTime = high_resolution_clock::now();
        queue.removeLast();
        endTime = high_resolution_clock::now();
        allTimes[4][run] = duration_cast<nanoseconds>(endTime - startTime).count();
        
        // Testa remove (middle)
        startTime = high_resolution_clock::now();
        queue.remove(middle);
        endTime = high_resolution_clock::now();
        allTimes[5][run] = duration_cast<nanoseconds>(endTime - startTime).count();
        
        // Testa getFirst
        startTime = high_resolution_clock::now();
        queue.getFirst();
        endTime = high_resolution_clock::now();
        allTimes[6][run] = duration_cast<nanoseconds>(endTime - startTime).count();
        
        // Testa getLast
        startTime = high_resolution_clock::now();
        queue.getLast();
        endTime = high_resolution_clock::now();
        allTimes[7][run] = duration_cast<nanoseconds>(endTime - startTime).count();
        
        // Testa get (middle)
        startTime = high_resolution_clock::now();
        queue.get(middle);
        endTime = high_resolution_clock::now();
        allTimes[8][run] = duration_cast<nanoseconds>(endTime - startTime).count();
    }

    // Calcula e escreve a mediana para cada método
    for (size_t method = 0; method < METHOD_NAMES.size(); method++) {
        sort(allTimes[method].begin(), allTimes[method].end());
        long medianTime = allTimes[method][RUNS / 2];
        
        writers[method] << "queue-cpp " << medianTime << " " << length << endl;
        writers[method].flush();
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Uso: " << argv[0] << " <arquivo_entrada>" << endl;
        return 1;
    }

    string inputFile = argv[1];
    string outputDir = "data/results/time/";

    // Abre arquivos de saída para cada método
    ofstream writers[METHOD_NAMES.size()];
    for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
        string outputFile = outputDir + METHOD_NAMES[i] + ".data";
        writers[i].open(outputFile, ios::app); // Modo append
        if (!writers[i]) {
            cerr << "Erro ao abrir arquivo de saída: " << outputFile << endl;
            return 1;
        }
    }

    try {
        ifstream reader(inputFile);
        if (!reader) {
            throw runtime_error("Não foi possível abrir o arquivo de entrada: " + inputFile);
        }

        string line;
        while (getline(reader, line)) {
            if (line.empty()) continue;
            
            // Converte a linha para vetor de inteiros
            vector<int> elements;
            size_t pos = 0;
            while ((pos = line.find(' ')) != string::npos) {
                string token = line.substr(0, pos);
                elements.push_back(stoi(token));
                line.erase(0, pos + 1);
            }
            if (!line.empty()) {
                elements.push_back(stoi(line));
            }

            testQueueOperations(elements, writers);
        }

        // Fecha todos os arquivos
        for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
            writers[i].close();
        }

    } catch (const exception& e) {
        cerr << "Erro: " << e.what() << endl;
        for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
            if (writers[i].is_open()) {
                writers[i].close();
            }
        }
        return 1;
    }

    return 0;
}
